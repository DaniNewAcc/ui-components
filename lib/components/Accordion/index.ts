export { default, useAccordionContext, useAccordionItemContext } from './Accordion';
export type {
  AccordionContentProps,
  AccordionContextProps,
  AccordionItemContextProps,
  AccordionItemProps,
  AccordionProps,
  AccordionTriggerProps,
} from './Accordion';
