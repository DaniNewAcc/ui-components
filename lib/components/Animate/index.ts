export { default } from './Animate';
export type { AnimateMethods, AnimateProps } from './Animate';
