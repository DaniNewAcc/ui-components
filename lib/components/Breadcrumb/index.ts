export { default, useBreadcrumbContext } from './Breadcrumb';
export type {
  BreadcrumbContextProps,
  BreadcrumbItemProps,
  BreadcrumbLinkProps,
  BreadcrumbProps,
  BreadcrumbSeparatorProps,
} from './Breadcrumb';
