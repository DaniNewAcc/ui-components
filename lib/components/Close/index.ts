export { default } from './Close';
export type { CloseProps } from './Close';
