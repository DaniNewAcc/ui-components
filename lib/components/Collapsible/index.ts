export { default, useCollapsibleContext, useCollapsibleItemContext } from './Collapsible';
export type {
  CollapsibleContentProps,
  CollapsibleContextProps,
  CollapsibleItemContextProps,
  CollapsibleItemProps,
  CollapsibleProps,
  CollapsibleTriggerProps,
} from './Collapsible';
