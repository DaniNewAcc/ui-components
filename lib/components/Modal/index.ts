export { default, useModalContext } from './Modal';
export type {
  ModalCloseProps,
  ModalContentProps,
  ModalContextProps,
  ModalDescriptionProps,
  ModalFooterProps,
  ModalOverlayProps,
  ModalPortalProps,
  ModalProps,
  ModalTitleProps,
  ModalTriggerProps,
} from './Modal';
