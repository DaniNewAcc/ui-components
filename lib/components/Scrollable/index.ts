export { default } from './Scrollable';
export type { ScrollableProps } from './Scrollable';
