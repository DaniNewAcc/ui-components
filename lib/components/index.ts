export {
  Accordion,
  AccordionContent,
  AccordionItem,
  AccordionTrigger,
} from '@components/Accordion';
export { default as Animate } from '@components/Animate';
export { default as Button } from '@components/Button';
export { default as Card } from '@components/Card';
export { default as Flex } from '@components/Flex';
export { default as Footer } from '@components/Footer';
export { default as Grid } from '@components/Grid';
export { default as Header } from '@components/Header';
export { default as Image } from '@components/Image';
export { default as Input } from '@components/Input';
export { default as Label } from '@components/Label';
export { default as Layout } from '@components/Layout';
export { default as Loader } from '@components/Loader';
export { default as Scrollable } from '@components/Scrollable';
export { Select, SelectDropdown, SelectOption, SelectTrigger } from '@components/Select';
export { default as Separator } from '@components/Separator';
export { Tabs, TabsContent, TabsList, TabsTrigger } from '@components/Tabs';
export { default as Text } from '@components/Text';
