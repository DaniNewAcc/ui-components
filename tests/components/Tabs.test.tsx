import { Tabs, TabsContent, TabsList, TabsTrigger } from '@/components';
import { cleanup, fireEvent, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

const renderTabs = (
  count = 3,
  defaultValue = 1,
  hasPadding = false,
  orientation: 'horizontal' | 'vertical' = 'horizontal',
  disabledTabs: boolean[] = []
) => {
  return render(
    <Tabs
      defaultValue={defaultValue}
      hasPadding={hasPadding}
      orientation={orientation}
      testId="tabsgroup"
    >
      <TabsList testId="tabslist">
        {Array.from({ length: count }, (_, i) => (
          <TabsTrigger key={i} value={i + 1} disabled={disabledTabs[i]}>
            Trigger {i + 1}
          </TabsTrigger>
        ))}
      </TabsList>
      {Array.from({ length: count }, (_, i) => (
        <TabsContent key={i} value={i + 1}>
          Content {i + 1}
        </TabsContent>
      ))}
    </Tabs>
  );
};

afterEach(cleanup);

describe('Tabs', () => {
  describe('Context behavior', () => {
    it('should throw an error when components are not wrapped into tabs', () => {
      expect(() =>
        render(
          <>
            <TabsList>
              <TabsTrigger value={1}>Unwrapped Trigger</TabsTrigger>
            </TabsList>
            <TabsContent value={1}>Unwrapped Content</TabsContent>
          </>
        )
      ).toThrowError('Tabs components need to be wrapped into <Tabs>.');
    });
  });

  describe('Rendering', () => {
    it('should render with padding', () => {
      const { getByTestId } = renderTabs(1, 1, true);
      expect(getByTestId('tabsgroup')).toHaveClass('ui:p-4');
    });

    it('should render required elements', () => {
      renderTabs(2);
      expect(screen.getByRole('tablist')).toBeInTheDocument();
      expect(screen.getAllByRole('tab')).toHaveLength(2);
      expect(screen.getByRole('tabpanel')).toHaveTextContent('Content 1');
    });

    it('should match snapshot', () => {
      const { container } = renderTabs(2);
      expect(container).toMatchSnapshot();
    });
  });

  describe('Disabled Tabs', () => {
    it('should skip disabled tabs when using keyboard navigation', async () => {
      renderTabs(3, 1, false, 'horizontal', [false, true, false]);

      const tabs = screen.getAllByRole('tab');
      const user = userEvent.setup();

      tabs[0].focus();

      await user.keyboard('{ArrowRight}');

      expect(tabs[2]).toHaveFocus();

      await user.keyboard('{ArrowLeft}');
      expect(tabs[0]).toHaveFocus();
    });
  });

  describe('Interaction', () => {
    beforeEach(() => {
      renderTabs(2);
    });

    it('should switch content on tab trigger click', () => {
      const tabsTrigger = screen.getByRole('tab', { name: 'Trigger 2' });

      fireEvent.click(tabsTrigger);
      expect(screen.getByRole('tabpanel')).toHaveTextContent('Content 2');
      expect(tabsTrigger).toHaveAttribute('aria-selected', 'true');
    });

    it('should not render non-active tab content', () => {
      cleanup();
      renderTabs(2, 1);
      expect(screen.queryByText('Content 2')).not.toBeInTheDocument();
    });

    it('should set focus visible styling on mousedown event', () => {
      const tabsTrigger = screen.getByRole('tab', { name: 'Trigger 2' });
      expect(tabsTrigger).not.toHaveClass('focus-visible');

      fireEvent.mouseDown(tabsTrigger);
      expect(tabsTrigger).not.toHaveClass('focus-visible');
    });
  });

  describe('Keyboard Navigation', () => {
    beforeEach(() => {
      renderTabs();
    });

    it('should move focus to the first tab when Tab key is pressed before Home key', async () => {
      const user = userEvent.setup();
      const tabs = screen.getAllByRole('tab');
      expect(tabs[0]).not.toHaveFocus();

      await user.keyboard('{Tab}');

      await user.keyboard('{Home}');

      await waitFor(() => {
        expect(tabs[0]).toHaveFocus();
      });
    });

    it('should move focus to the last tab when End key is pressed', async () => {
      const user = userEvent.setup();
      const tabs = screen.getAllByRole('tab');

      tabs[0].focus();
      await user.keyboard('{End}');
      await waitFor(() => {
        expect(tabs[tabs.length - 1]).toHaveFocus();
      });
    });

    it('should move focus to the tabpanel related to tabtrigger focused when Tab is pressed', async () => {
      const user = userEvent.setup();
      const tabs = screen.getAllByRole('tab');
      const panels = screen.getAllByRole('tabpanel');

      tabs[0].focus();

      await waitFor(() => {
        expect(tabs[0]).toHaveFocus();
      });

      await user.keyboard('{Tab}');

      panels[0].focus();

      await waitFor(() => {
        expect(panels[0]).toHaveFocus();
      });
    });

    it('should move focus with left and right arrow keys', async () => {
      const user = userEvent.setup();
      const tabs = screen.getAllByRole('tab');

      tabs[0].focus();
      await waitFor(() => expect(tabs[0]).toHaveFocus());

      await user.keyboard('{ArrowRight}');
      await waitFor(() => expect(tabs[1]).toHaveFocus());

      await user.keyboard('{ArrowRight}');
      await waitFor(() => expect(tabs[2]).toHaveFocus());

      await user.keyboard('{ArrowRight}');
      await waitFor(() => expect(tabs[0]).toHaveFocus());

      await user.keyboard('{ArrowLeft}');
      await waitFor(() => expect(tabs[2]).toHaveFocus());
    });

    it('should move focus with up and down arrow keys when vertical orientation is set', async () => {
      cleanup();
      renderTabs(3, 1, false, 'vertical');
      const tabs = screen.getAllByRole('tab');
      const user = userEvent.setup();

      const tabList = screen.getByRole('tablist');
      expect(tabList).toHaveAttribute('aria-orientation', 'vertical');

      tabs[0].focus();
      await waitFor(() => expect(tabs[0]).toHaveFocus());

      await user.keyboard('{ArrowDown}');
      await waitFor(() => expect(tabs[1]).toHaveFocus());

      await user.keyboard('{ArrowDown}');
      await waitFor(() => expect(tabs[2]).toHaveFocus());

      await user.keyboard('{ArrowDown}');
      await waitFor(() => expect(tabs[0]).toHaveFocus());

      await user.keyboard('{ArrowUp}');
      await waitFor(() => expect(tabs[2]).toHaveFocus());
    });
  });
});
