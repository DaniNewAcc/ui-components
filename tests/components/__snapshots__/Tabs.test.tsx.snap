// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tabs > Rendering > should match snapshot 1`] = `
<div>
  <div
    class="ui:flex ui:w-fit ui:justify-between ui:gap-4 ui:rounded-md ui:shadow-md ui:self-stretch ui:flex-col"
    data-testid="tabsgroup"
  >
    <div
      aria-activedescendant="tab-1"
      aria-label="Tabs"
      aria-orientation="horizontal"
      class="ui:inline-flex ui:w-fit ui:shrink-0 ui:items-center ui:justify-center ui:rounded-t-md ui:bg-transparent ui:flex-row"
      data-testid="tabslist"
      role="tablist"
    >
      <button
        aria-controls="content-1"
        aria-selected="true"
        class="ui:inline-flex ui:gap-2 ui:items-center ui:justify-center ui:font-semibold ui:border ui:relative ui:shadow-sm ui:whitespace-nowrap ui:cursor-pointer ui:transition-all ui:duration-200 ui:active:shadow-none ui:focus:transition-none ui:focus:duration-initial ui:focus:outline-hidden ui:disabled:pointer-events-none ui:disabled:opacity-50 ui:disabled:shadow-none ui:border-transparent ui:px-2 ui:py-1 ui:text-sm ui:bg-primary-50 ui:text-primary-600 ui:focus-visible:ring-0 ui:last:rounded-tr-sm ui:flex-1 ui:first:rounded-tl-sm"
        id="trigger-1"
        role="tab"
        tabindex="0"
        type="button"
      >
        <span
          class=""
        >
          Trigger 
          1
        </span>
      </button>
      <button
        aria-controls="content-2"
        aria-selected="false"
        class="ui:inline-flex ui:gap-2 ui:items-center ui:justify-center ui:font-semibold ui:border ui:relative ui:shadow-sm ui:whitespace-nowrap ui:cursor-pointer ui:transition-all ui:duration-200 ui:active:shadow-none ui:focus:transition-none ui:focus:duration-initial ui:focus:outline-hidden ui:disabled:pointer-events-none ui:disabled:opacity-50 ui:disabled:shadow-none ui:border-transparent ui:px-2 ui:py-1 ui:text-sm ui:focus-visible:ring-0 ui:last:rounded-tr-sm ui:flex-1 ui:first:rounded-tl-sm"
        id="trigger-2"
        role="tab"
        tabindex="-1"
        type="button"
      >
        <span
          class=""
        >
          Trigger 
          2
        </span>
      </button>
    </div>
    <div
      aria-labelledby="trigger-1"
      class="ui:flex-1 ui:overflow-y-auto"
      id="panel-1"
      role="tabpanel"
      tabindex="0"
    >
      Content 
      1
    </div>
  </div>
</div>
`;
